rust/main/chains/hyperlane-ethereum/src/signer/yubi.rs

```diff
@@ -0,0 +1,152 @@
+use ethers_signers::Wallet;
+use signature::digest::Digest;
+use signature::digest::FixedOutput;
+use signature::{DigestSigner, Error};
+use std::collections::HashMap;
+use std::hash::{Hash, Hasher};
+use std::sync::Arc;
+use std::sync::{LazyLock, RwLock};
+use yubihsm::ecdsa::sec1::FromEncodedPoint;
+use yubihsm::ecdsa::sec1::ToEncodedPoint;
+use yubihsm::object;
+use yubihsm::Client;
+use yubihsm::HttpConfig;
+use yubihsm::{ecdsa::Signer as YubiSigner, Credentials};
+
+/// Unique key for a yubihsm client in the global cache.
+/// The global caches is used so that we can reuse yubihsm sessions across threads, preventing resource exhaustion.
+struct YubihsmClientCacheKey {
+    config: HttpConfig,
+    credentials: Credentials,
+    id: object::Id,
+}
+
+impl PartialEq for YubihsmClientCacheKey {
+    fn eq(&self, other: &Self) -> bool {
+        self.config.addr == other.config.addr
+            && self.config.port == other.config.port
+            && self.config.timeout_ms == other.config.timeout_ms
+            && self.credentials.authentication_key_id == other.credentials.authentication_key_id
+            && self.id == other.id
+    }
+}
+
+impl Eq for YubihsmClientCacheKey {}
+
+impl Hash for YubihsmClientCacheKey {
+    fn hash<H: Hasher>(&self, state: &mut H) {
+        self.config.addr.hash(state);
+        self.config.port.hash(state);
+        self.config.timeout_ms.hash(state);
+        self.credentials.authentication_key_id.hash(state);
+        self.id.hash(state);
+    }
+}
+
+/// A thread safe cache for signers against a yubihsm.
+/// Yubihsms have a limited number of sessions (16), which remain open for at least 30 seconds. If a number of connections is requested
+/// too quickly, the devices resources will become exhausted leading to failures.
+static YUBIHSM_CLIENT_CACHE: LazyLock<RwLock<HashMap<YubihsmClientCacheKey, WrappedSigner>>> =
+    LazyLock::new(|| RwLock::new(HashMap::new()));
+
+/// Get a signer.
+/// If a matching signer is in the cache, the cached version is returned. Otherwise a new signer is initialized, and added to the cache.
+/// This function is thread safe.
+fn get_signer_with_cache(
+    http_config: HttpConfig,
+    credentials: Credentials,
+    id: object::Id,
+) -> WrappedSigner {
+    let cache_key = YubihsmClientCacheKey {
+        config: http_config.clone(),
+        credentials: credentials.clone(),
+        id: id.clone(),
+    };
+
+    // If a client already exists for this key, return a clone of the Arc.
+    {
+        let cache = YUBIHSM_CLIENT_CACHE.read().unwrap();
+
+        if let Some(existing_client) = cache.get(&cache_key) {
+            return Arc::clone(existing_client);
+        }
+    }
+
+    // Build a new client and add it to the cache.
+    let client = YubiHsmSigner::connect(http_config, credentials);
+    let signer = Arc::new(YubiSigner::create(client, id).unwrap());
+    {
+        let mut cache = YUBIHSM_CLIENT_CACHE.write().unwrap();
```

Comment:
[Original comment](https://github.com/Tessellated-io/hyperlane-monorepo/pull/2#discussion_r1979734193)

> https://github.com/Tessellated-io/hyperlane-monorepo/pull/2

Done. Also updated the `connect` function below from `unwrap` -> `expect`.